### PR TITLE
Client pipelines

### PR DESCRIPTION
**Pull request by davidfowl**
- Reworked the Client to be based on pipelines instead of Channels
- SendAsync no longer fails if the http request itself fails but the connection is closed as a result.
- Updated tests

### CODE REVIEW

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 1]
```diff
@@ -83,67 +81,67 @@ public Task StartAsync(Uri url, Channel<byte[], SendMessage> application, Transf
...
+                    _logger.ReceiveStopped();
+                }
             }
         }
 
         public async Task StopAsync()
         {
             _logger.TransportStopping();
             _transportCts.Cancel();
-            _application.Writer.TryComplete();
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Why is this gone?

File: test/Microsoft.AspNetCore.SignalR.Client.Tests/LongPollingTransportTests.cs [hunk 2]
```diff
@@ -166,13 +160,19 @@ public class LongPollingTransportTests
...
                     var exception =
-                        await Assert.ThrowsAsync<HttpRequestException>(async () => await transportToConnection.Reader.Completion.OrTimeout());
+                        await Assert.ThrowsAsync<HttpRequestException>(async () =>
+                        {
+                            async Task ReadAsync()
+                            {
+                                await pair.Transport.Input.ReadAsync();
+                            }
+
+                            await ReadAsync().OrTimeout();
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.SignalR.Client.Tests/LongPollingTransportTests.cs]: Looks like overkill o.O

File: src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs [hunk 3]
```diff
@@ -237,7 +236,8 @@ private Task StartAsyncCore()
                         // Suppress (but log) the exception, this is user code
                         _logger.ErrorDuringClosedEvent(ex);
                     }
-                });
+
+                }, null);
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: Yes but the code was always capturing. I didn't want the diff to be big so I left it as is. Will change it in another PR (lots of low hanging fruit 😄 )

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs [hunk 4]
```diff
@@ -122,17 +123,11 @@ public Task StartAsync(Uri url, Channel<byte[], SendMessage> application, Transf
...
                         if (payload.Length > 0)
                         {
-                            while (!_application.Writer.TryWrite(payload))
-                            {
-                                if (cancellationToken.IsCancellationRequested || !await _application.Writer.WaitToWriteAsync(cancellationToken))
-                                {
-                                    return;
-                                }
-                            }
+                            await _application.Output.WriteAsync(payload);
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: https://github.com/dotnet/corefx/pull/27007#discussion_r167422800

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs (hunk 1 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: It isn't required. All completion happens in here now https://github.com/aspnet/SignalR/pull/1435/files/c34463436737ee897219649a4e61c56e8f646346#diff-579fedec848db9fecbd95f172d7654e7R66. Both tasks should always end and we should clean up in the same place.

File: test/Microsoft.AspNetCore.SignalR.Client.Tests/LongPollingTransportTests.cs (hunk 2 above)
Comment (davidfowl) [reply test/Microsoft.AspNetCore.SignalR.Client.Tests/LongPollingTransportTests.cs]: This was just to get a `Task` from the pipe awaiter.

Comment (davidfowl) [reply test/Microsoft.AspNetCore.SignalR.Client.Tests/LongPollingTransportTests.cs]: I could use one of the helpers I used in the other tests.

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Encoders/Base64Encoder.cs [hunk 5]
```diff
@@ -10,16 +10,17 @@ namespace Microsoft.AspNetCore.SignalR.Internal.Encoders
 {
     public class Base64Encoder : IDataEncoder
     {
-        public ReadOnlySpan<byte> Decode(byte[] payload)
+        public bool TryDecode(ref ReadOnlySpan<byte> buffer, out ReadOnlySpan<byte> data)
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/Encoders/Base64Encoder.cs]: Add a test, and also fix the compilation errors :P

Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Common/Internal/Encoders/Base64Encoder.cs]: Yea doing this in VS code 😄. I'll spend more time looking at this soon.

**Comment by davidfowl** [comment]
ZOMG green!

**Comment by BrennanConroy** [comment]
But will it blend? I mean, will it pass if you run it again?

**Comment by davidfowl** [comment]
Lets see.

**Comment by davidfowl** [comment]
@BrennanConroy we need to add logging to a few key places to help diagnose issues:
- SSE Transport and LongPolling transport
- HubConnection needs some logs when it's processing messages (how many it parsed)
- Should also show print the message when tracing.

**Comment by BrennanConroy** [comment]
I also found an issue in a couple of our logs at Trace level

**Comment by davidfowl** [comment]
Argggg, OSX!!! 😄

**Comment by davidfowl** [comment]
This test, failed twice now.

- ConnectionCanSendAndReceiveDifferentMessageSizesWebSocketsTransport

@BrennanConroy can we get artifacts from travis?

**Comment by BrennanConroy** [comment]
Nope, they only support pushing to an AWS subscription you own

**Comment by davidfowl** [comment]
Well I'm going to give this a few more goes then rage merge. Things need to be cleaned up. All of the parsers need to be rewritten.

**Pull request merged**